gax/src/main/java/com/google/api/gax/retrying/RetryAlgorithm.java

```diff
@@ -66,63 +95,170 @@ public RetryAlgorithm(
+      RetryingContext context,
+      Throwable previousThrowable,
+      ResponseT previousResponse,
+      TimedAttemptSettings previousSettings) {
+    // a small optimization that avoids calling relatively heavy methods
     // like timedAlgorithm.createNextAttempt(), when it is not necessary.
-    if (!resultAlgorithm.shouldRetry(prevThrowable, prevResponse)) {
+    if (!shouldRetryBasedOnResult(context, previousThrowable, previousResponse)) {
       return null;
     }
 
     TimedAttemptSettings newSettings =
-        resultAlgorithm.createNextAttempt(prevThrowable, prevResponse, prevSettings);
+        createNextAttemptBasedOnResult(
+            context, previousThrowable, previousResponse, previousSettings);
     if (newSettings == null) {
-      newSettings = timedAlgorithm.createNextAttempt(prevSettings);
+      newSettings = createNextAttemptBasedOnTiming(context, previousSettings);
     }
     return newSettings;
   }
 
+  @SuppressWarnings("deprecation")
+  private TimedAttemptSettings createNextAttemptBasedOnResult(
+      RetryingContext context,
+      Throwable previousThrowable,
+      ResponseT previousResponse,
+      TimedAttemptSettings previousSettings) {
+    if (resultAlgorithmWithContext != null && context != null) {
+      return resultAlgorithmWithContext.createNextAttempt(
+          context, previousThrowable, previousResponse, previousSettings);
+    }
+    return getResultAlgorithm()
+        .createNextAttempt(previousThrowable, previousResponse, previousSettings);
+  }
+
+  @SuppressWarnings("deprecation")
+  private TimedAttemptSettings createNextAttemptBasedOnTiming(
+      RetryingContext context, TimedAttemptSettings previousSettings) {
+    if (timedAlgorithmWithContext != null && context != null) {
+      return timedAlgorithmWithContext.createNextAttempt(context, previousSettings);
+    }
+    return getTimedAlgorithm().createNextAttempt(previousSettings);
+  }
+
   /**
    * Returns {@code true} if another attempt should be made, or {@code false} otherwise.
    *
-   * @param prevThrowable exception thrown by the previous attempt or null if a result was returned
-   *     instead
-   * @param prevResponse response returned by the previous attempt or null if an exception was
+   * @param previousThrowable exception thrown by the previous attempt or null if a result was
+   *     returned instead
+   * @param previousResponse response returned by the previous attempt or null if an exception was
    *     thrown instead
    * @param nextAttemptSettings attempt settings, which will be used for the next attempt, if
    *     accepted
    * @throws CancellationException if the retrying process should be canceled
    * @return {@code true} if another attempt should be made, or {@code false} otherwise
    */
   public boolean shouldRetry(
-      Throwable prevThrowable, ResponseT prevResponse, TimedAttemptSettings nextAttemptSettings)
+      Throwable previousThrowable,
```

Comment:
Sorry for confusing with my `@Deprecate` request. I think we should just mark the methods and constructors in `RetryAlgorithm` class as deprecated (the old ones) and make sure that the non-deprecated methos (new ones) are called by the deprecated methods (which is what is happening now, so it is all good), otherwise we will be using our own deprecated methods to implement non-deprecated methods, which is wrong.

For the rest of it, please undeprecate the methods back, as is is clear that `RetryAlgorithm` cannot be used in both contexts (as context-aware and as unaware), but the actual interfaces (timed and result) kind of still potentially can have meaningful implementation for all the methods, but most importantly it is too much deprecation happening on interface level, I'm afraid we are going to get too many deprecaton warnings.

So to summarize: deprecate the old constructor and methods in `RetryAlgorithm` class, undeprecate the other stuff (the TimedRetryAlgorithm and ResultRetryAlgorithm  methods - in both places, the `@Deprecated` annotation and `@deprecated` java doc element, maybe adding in the general java doc section of each element a notion that the other method should be preferred).

Reply:
Done